crates/roc_std/src/roc_dict.rs

```diff
@@ -1,31 +1,41 @@
-use crate::roc_list::{self, RocList};
+use crate::roc_list::RocList;
 use core::{
     fmt::{self, Debug},
     hash::Hash,
 };
 
 #[derive(Default, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct RocDict<K, V>(RocList<(K, V)>);
+pub struct RocDict<K, V>(RocList<RocDictItem<K, V>>);
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct RocDictItem<K, V> {
+    key: K,
+    value: V,
+}
```

Comment:
something I just thought of: roc sorts struct fields by alignment, so we don't guarantee that the key comes first in memory. That means that this implementation of `RocDict` isn't entirely faithful, and might be dangerous when used for FFI.

I don't think there is a great solution for this either; in rust, types cannot depend on values (yet? who knows when rust will be dependently typed)

Reply:
hmm… when would that come up? Are we creating RocDictItems as if they were Roc structs somewhere?

Reply:
On the Roc side, `RocDict` is currently implemented as a list of key-value pairs. So if the value has a larger alignment than the key, the pairs will actually be value-key pairs.

We could do something like

```
struct KeyFirst<K,V> {
  key: K,
  value: V
}

struct ValueFirst<K,V> {
  value: V,
  key: K
}

union RocDictItem<K, V> {
    key_first: KeyFirst<K, V>,
    value_first: ValueFirst<K, V> ,
}

impl<K,V> RocDictItem {
  fn key(&self) {
    if std::mem::align_of::<K>() >= std::mem::align_of::<V> {
      self.key_first.key
    } else {
      self.value_first.key
    }
  }
}
```

Reply:
oh, slick. Gonna get something like that in, thanks for the pointer.